cpp/src/reductions/minmax.cu

```diff
@@ -0,0 +1,183 @@
+    : min_val(min_val_), max_val(max_val_), min_valid(min_valid_), max_valid(max_valid_){};
+};
+
+/**
+ * @brief functor that accepts two minmax_pairs and returns a
+ * minmax_pair whose minimum and maximum values are the min() and max()
+ * respectively of the minimums and maximums of the input pairs. Respects
+ * validity.
+ *
+ */
+template <typename T>
+struct minmax_with_null_binary_op
+  : public thrust::binary_function<minmax_pair<T>, minmax_pair<T>, minmax_pair<T>> {
+  __host__ __device__ minmax_pair<T> operator()(const minmax_pair<T> &x,
+                                                const minmax_pair<T> &y) const
+  {
+    T x_min = x.min_valid ? x.min_val : cudf::DeviceMin::identity<T>();
+    T y_min = y.min_valid ? y.min_val : cudf::DeviceMin::identity<T>();
+    T x_max = x.max_valid ? x.max_val : cudf::DeviceMax::identity<T>();
+    T y_max = y.max_valid ? y.max_val : cudf::DeviceMax::identity<T>();
+
+    // The only invalid situation is if we compare two invalid values.
+    // Otherwise, we are certain to select a valid value due to the
+    // identity functions above changing the comparison value.
+    bool valid_min_result = x.min_valid || y.min_valid;
+    bool valid_max_result = x.max_valid || y.max_valid;
+
+    return minmax_pair<T>{
+      thrust::min(x_min, y_min), valid_min_result, thrust::max(x_max, y_max), valid_max_result};
+  }
+};
+
+/**
+ * @brief functor that accepts two minmax_pairs and returns a
+ * minmax_pair whose minimum and maximum values are the min() and max()
+ * respectively of the minimums and maximums of the input pairs. Expects
+ * no null values.
+ *
+ */
+template <typename T>
+struct minmax_no_null_binary_op
+  : public thrust::binary_function<minmax_pair<T>, minmax_pair<T>, minmax_pair<T>> {
+  __host__ __device__ minmax_pair<T> operator()(const minmax_pair<T> &x,
+                                                const minmax_pair<T> &y) const
+  {
+    return minmax_pair<T>{
+      thrust::min(x.min_val, y.min_val), true, thrust::max(x.max_val, y.max_val), true};
+  }
+};
+
+/**
+ * @brief functor that calls thrust::transform_reduce to produce a std::pair
+ * of scalars that represent the minimum and maximum values of the input data
+ * respectively. Note that dictionaries and non-relationally comparable objects
+ * are not supported.
+ *
+ */
+struct minmax_functor {
+  template <typename T>
+  // unable to support fixed point due to DeviceMin/DeviceMax not supporting fixed point
+  std::enable_if_t<cudf::is_relationally_comparable<T, T>() and
+                     not std::is_same<T, dictionary32>::value and not cudf::is_fixed_point<T>(),
+                   std::pair<std::unique_ptr<scalar>, std::unique_ptr<scalar>>>
+  operator()(const cudf::column_view &col, rmm::mr::device_memory_resource *mr, cudaStream_t stream)
+  {
+    auto device_col = column_device_view::create(col, stream);
+
+    // compute minimum and maximum values
+    minmax_pair<T> result;
+    if (col.nullable()) {
```

Comment:
```suggestion
    if (col.has_nulls()) {
```

Reply:
I was worried about the performance here of marching all the data to see if there are nulls. Is it generally better to do that and optimize the work or risk running the null version of the code when it isn't necessary?

Reply:
My general philosophy is that you should assume `has_nulls` has it's value already present. It's going to need to be computed at some point in the lifetime of a column, but once it's computed, it will stay computed. If that point so happens to be inside your function, then so be it.